install/cupy_builder/cupy_setup_build.py

```diff
@@ -336,21 +336,35 @@ def make_extensions(ctx: Context, compiler, use_cython):
             os.symlink(f'{os.environ["BUILD_PREFIX"]}/x86_64-conda-linux-gnu/'
                        'bin/x86_64-conda-linux-gnu-ld',
                        f'{os.environ["BUILD_PREFIX"]}/bin/ld')
-        if (int(os.environ.get('CONDA_BUILD_CROSS_COMPILATION', 0)) == 1 or
-                os.environ.get('CONDA_OVERRIDE_CUDA', '0').startswith('12')):
+        if (PLATFORM_LINUX and (
+                int(os.environ.get('CONDA_BUILD_CROSS_COMPILATION', 0)) == 1 or
+                os.environ.get('CONDA_OVERRIDE_CUDA', '0').startswith('12'))):
             # If cross-compiling, we need build_and_run() & build_shlib() to
             # use the compiler on the build platform to generate stub files
             # that are executable in the build environment, not the target
             # environment.
             compiler = ccompiler.new_compiler()
-            cc = os.environ['CC_FOR_BUILD' if PLATFORM_LINUX else 'CC']
-            cxx = os.environ['CXX_FOR_BUILD' if PLATFORM_LINUX else 'CXX']
+            sysconfig.customize_compiler(compiler)
+            # Need to match and replace these
+            # https://github.com/pypa/distutils/blob/30b7331b07fbc404959cb37ac311afdfb90813be/distutils/unixccompiler.py#L117-L129
+            cc = os.environ['CC_FOR_BUILD']
+            cxx = os.environ['CXX_FOR_BUILD']
+            ar = os.environ['BUILD'] + "-ar"
+            compiler.preprocessor = None
             compiler.compiler = [cc,]
-            compiler.compiler_cxx = [cxx,]
             compiler.compiler_so = [cc,]
-            compiler.linker_exe = [cc, f'-B{os.environ["BUILD_PREFIX"]}/bin']
+            compiler.compiler_cxx = [cxx,]
+            compiler.compiler_so_cxx = [cxx,]
```

Comment:
@leofang please see this line specifically. Without this we get the error

```python
command 'c++' failed: No such file or directory
```

Previously Distutils was part of CPython as of 3.11. When it was, this was [not defined]( https://github.com/python/cpython/blob/795f2597a4be988e2bb19b69ff9958e981cb894e/Lib/distutils/unixccompiler.py#L55-L63 )

However Python 3.12 removed Distutils. Please see [PEP 632]( https://peps.python.org/pep-0632 )

Think we mostly don't notice as [Setuptools ships its own Distutils]( https://setuptools.pypa.io/en/latest/deprecated/distutils-legacy.html ), which it uses by default. We probably should move to it explicitly in CuPy as Distutils (even within Setuptools) [is deprecated]( https://setuptools.pypa.io/en/latest/deprecated/distutils-legacy.html#prefer-setuptools )

Since Setuptools has taken over Distutils support, they have been making changes like the C++ changes we adapted here, which were included in PR: https://github.com/pypa/distutils/pull/228

Now why did this only affect us now? Think Setuptools updated its vendored Distutils recently ( like commit: https://github.com/pypa/setuptools/commit/2c937116cc0dcd9b26b6070e89a3dc5dcbedc2ae ) and this went out [72.2.0]( https://github.com/pypa/setuptools/releases/tag/v72.2.0 ), which conda-forge dutifully packaged recently ( https://github.com/conda-forge/setuptools-feedstock/pull/355 )

Reply:
Sorry I edited that message a few times to fill in more details. Hopefully that helps clarify. Please ask more questions as needed